File: Hammerspoon 2/Modules/hs.hotkey/HSHotkey.swift
//
//  HSHotkey.swift
//  Hammerspoon 2
//

import Foundation
import JavaScriptCore
import AppKit
import Carbon

// MARK: - Protocol

/// Object representing a system-wide hotkey. You should not create these objects directly, but rather, use the methods in hs.hotkey to instantiate these.
@objc protocol HSHotkeyAPI: HSTypeAPI, JSExport {
    /// {string[]} The modifier keys this hotkey was bound with, as originally passed to bind()/create()
    /// - Example:
    /// ```js
    /// const hk = hs.hotkey.bind(["cmd"], "h", () => {})
    /// console.log(hk.mods)
    /// ```
    @objc var mods: [String] { get }

    /// {string} The key this hotkey was bound with, as originally passed to bind()/create()
    /// - Example:
    /// ```js
    /// const hk = hs.hotkey.bind(["cmd"], "h", () => {})
    /// console.log(hk.key)
    /// ```
    @objc var key: String { get }

    /// {string | null} An optional description of what this hotkey does, or null if none was set.
    /// When set, it is shown as an on-screen toast via `hs.ui.alert()` (duration controlled by
    /// `hs.hotkey.alertDuration`) just before the hotkey's callback runs: before the pressed
    /// callback if one exists, otherwise before the released callback if one exists.
    /// - Example:
    /// ```js
    /// const hk = hs.hotkey.bind(["cmd"], "h", () => {})
    /// hk.message = "Say hello"
    /// ```
    @objc var message: String? { get set }

    /// {(() => void) | null} The callback function to be called repeatedly while the hotkey is held down, or null to remove it. Repeats at the system keyboard-repeat delay/interval, matching how held-down keys repeat elsewhere in macOS.
    /// - Example:
    /// ```js
    /// const hk = hs.hotkey.bind(["cmd"], "h", () => {})
    /// hk.callbackRepeat = () => console.log("still held")
    /// ```
    @objc var callbackRepeat: JSFunction? { get set }

    /// Enable the hotkey
    /// - Returns: True if the hotkey was enabled, otherwise False
    /// - Example:
    /// ```js
    /// const hk = hs.hotkey.bind(["cmd"], "h", () => {})
    /// hk.enable()
    /// ```
    @objc func enable() -> Bool

    /// Disable the hotkey
    /// - Example:
    /// ```js
    /// const hk = hs.hotkey.bind(["cmd"], "h", () => {})
    /// hk.disable()
    /// ```
    @objc func disable()

    /// Check if the hotkey is currently enabled
    /// - Returns: True if the hotkey is enabled, otherwise False
    /// - Example:
    /// ```js
    /// const hk = hs.hotkey.bind(["cmd"], "h", () => {})
    /// console.log(hk.isEnabled())
    /// ```
    @objc func isEnabled() -> Bool

    /// {(() => void) | null} The callback function to be called when the hotkey is pressed, or null to remove it
    /// - Example:
    /// ```js
    /// const hk = hs.hotkey.bind(["cmd"], "h", () => {})
    /// hk.callbackPressed = () => console.log("new handler")
    /// ```
    @objc var callbackPressed: JSFunction? { get set }

    /// {(() => void) | null} The callback function to be called when the hotkey is released, or null to remove it
    /// - Example:
    /// ```js
    /// const hk = hs.hotkey.bind(["cmd"], "h", () => {})
    /// hk.callbackReleased = () => console.log("released")
    /// ```
    @objc var callbackReleased: JSFunction? { get set }

    /// Disable and permanently remove this hotkey, releasing all associated resources
    /// - Example:
    /// ```js
    /// const hk = hs.hotkey.bind(["cmd"], "h", () => {})
    /// hk.destroy()
    /// ```
    @objc func destroy()
}

// MARK: - Implementation

@_documentation(visibility: private)
@MainActor
@safe
@objc class HSHotkey: NSObject, HSHotkeyAPI {
    @objc var typeName = "HSHotkey"

    @objc func toString() -> String {
        return "<\(typeName): keyCode \(keyCode), modifiers \(modifiers)>"
    }

    nonisolated override var description: String {
        MainActor.assumeIsolated { toString() }
    }
    private let keyCode: UInt32
    private let modifiers: UInt32
    @objc let mods: [String]
    @objc let key: String
    @objc var message: String?
    private var _callbackPressed: JSCallback?
    private var _callbackReleased: JSCallback?
    private var _callbackRepeat: JSCallback?

    @objc var callbackPressed: JSFunction? {
        get { _callbackPressed?.value }
        set {
            _callbackPressed?.detach(from: self)
            _callbackPressed = newValue.flatMap { JSCallback(value: $0, owner: self, silentOnUndefined: true) }
        }
    }
    @objc var callbackReleased: JSFunction? {
        get { _callbackReleased?.value }
        set {
            _callbackReleased?.detach(from: self)
            _callbackReleased = newValue.flatMap { JSCallback(value: $0, owner: self, silentOnUndefined: true) }
        }
    }
    @objc var callbackRepeat: JSFunction? {
        get { _callbackRepeat?.value }
        set {
            _callbackRepeat?.detach(from: self)
            _callbackRepeat = newValue.flatMap { JSCallback(value: $0, owner: self, silentOnUndefined: true) }
        }
    }

    nonisolated(unsafe) private var carbonHotKeyRef: EventHotKeyRef?
    private var enabled = false
    private let hotkeyID: UInt32
    private var repeatDelayTimer: Timer?
    private var repeatIntervalTimer: Timer?

    init(keyCode: UInt32, modifiers: UInt32, mods: [String], key: String,
         callbackPressed: JSFunction? = nil, callbackReleased: JSFunction? = nil) {
        self.keyCode = keyCode
        self.modifiers = modifiers
        self.mods = mods
        self.key = key
        self.hotkeyID = HotkeyManager.shared.nextID
        super.init()

        self.callbackPressed = callbackPressed
        self.callbackReleased = callbackReleased
    }

    isolated deinit {
        destroy()
        AKGarbage("deinit of HSHotkey: id=\(hotkeyID)")
    }

    /// Returns true if this hotkey was bound with the given key code and modifier flags.
    func matches(keyCode: UInt32, modifiers: UInt32) -> Bool {
        self.keyCode == keyCode && self.modifiers == modifiers
    }

    @objc func destroy() {
        disable()
        _callbackPressed?.detach(from: self)
        _callbackPressed = nil
        _callbackReleased?.detach(from: self)
        _callbackReleased = nil
        _callbackRepeat?.detach(from: self)
        _callbackRepeat = nil
    }

    @objc func enable() -> Bool {
        guard !enabled else { return true }

        let hotKeyID = EventHotKeyID(
            signature: OSType(("HMSP" as NSString).fourCharCode),
            id: hotkeyID
        )
        let status = unsafe RegisterEventHotKey(
            keyCode, modifiers, hotKeyID,
            GetEventDispatcherTarget(), 0, &carbonHotKeyRef
        )

        if status != noErr {
            AKError("hs.hotkey: Failed to register hotkey (error \(status))")
            return false
        }

        enabled = true
        HotkeyManager.shared.register(hotkeyID: hotkeyID, hotkey: self)
        return true
    }

    @objc func disable() {
        stopRepeating()
        guard enabled, let ref = unsafe carbonHotKeyRef else { return }
        unsafe UnregisterEventHotKey(ref)
        unsafe carbonHotKeyRef = nil
        HotkeyManager.shared.unregister(hotkeyID: hotkeyID)
        enabled = false
    }

    @objc func isEnabled() -> Bool { enabled }

    func trigger(eventKind: UInt32) {
        switch eventKind {
        case UInt32(kEventHotKeyPressed):
            // A message is shown before the pressed callback whenever one exists,
            // whether or not a released callback also exists.
            if let context = _callbackPressed?.value?.context {
                showMessageAlert(in: context)
            }
            invoke(_callbackPressed?.value)
            startRepeatingIfNeeded()
        case UInt32(kEventHotKeyReleased):
            // Only shown before the released callback when there's no pressed callback
            // to have already shown it.
            if _callbackPressed == nil, let context = _callbackReleased?.value?.context {
                showMessageAlert(in: context)
            }
            invoke(_callbackReleased?.value)
            stopRepeating()
        default:
            AKError("hs.hotkey: Unknown event kind: \(eventKind)")
        }
    }

    private func invoke(_ callback: JSFunction?) {
        guard let callback, !callback.isNull else { return }
        callback.call(withArguments: [])
        if let context = callback.context,
           let exc = context.exception, !exc.isUndefined {
            AKError("hs.hotkey: Error in callback: \(exc.toString() ?? "unknown")")
            context.exception = nil
        }
    }

    // MARK: - Message alert

    private func showMessageAlert(in context: JSContext) {
        guard let message, !message.isEmpty else { return }
        // message is passed as a JSValue argument (not interpolated into evaluated source)
        // so arbitrary message content can never be interpreted as JavaScript.
        guard let hsUI = context.evaluateScript("hs.ui"), !hsUI.isUndefined else { return }
        guard let alert = hsUI.invokeMethod("alert", withArguments: [message]) else { return }
        let duration = context.evaluateScript("hs.hotkey.alertDuration")?.toDouble() ?? 1.0
        guard let sized = alert.invokeMethod("duration", withArguments: [duration]) else { return }
        sized.invokeMethod("show", withArguments: [])
    }

    // MARK: - Repeat

    private func startRepeatingIfNeeded() {
        guard _callbackRepeat != nil else { return }
        stopRepeating()
        repeatDelayTimer = Timer.scheduledTimer(withTimeInterval: NSEvent.keyRepeatDelay, repeats: false) { [weak self] _ in
            MainActor.assumeIsolated {
                self?.beginRepeating()
            }
        }
    }

    private func beginRepeating() {
        fireRepeatCallback()
        repeatIntervalTimer = Timer.scheduledTimer(withTimeInterval: NSEvent.keyRepeatInterval, repeats: true) { [weak self] _ in
            MainActor.assumeIsolated {
                self?.fireRepeatCallback()
            }
        }
    }

    private func fireRepeatCallback() {
        invoke(_callbackRepeat?.value)
    }

    private func stopRepeating() {
        repeatDelayTimer?.invalidate()
        repeatDelayTimer = nil
        repeatIntervalTimer?.invalidate()
        repeatIntervalTimer = nil
    }
}

// MARK: - Hotkey Manager

@_documentation(visibility: private)
@safe @MainActor
class HotkeyManager {
    static let shared = HotkeyManager()

    private var _nextID: UInt32 = 1
    var nextID: UInt32 {
        defer { _nextID += 1 }
        return _nextID
    }
    private var hotkeys: [UInt32: HSHotkey] = [:]
    nonisolated(unsafe) private var eventHandler: EventHandlerRef?
    nonisolated(unsafe) private var contextPtr: UnsafeMutablePointer<HotkeyManager>?

    private init() {
        setupEventHandler()
    }

    isolated deinit {
        if let handler = unsafe eventHandler {
            unsafe RemoveEventHandler(handler)
        }
        unsafe contextPtr?.deinitialize(count: 1)
        unsafe contextPtr?.deallocate()
    }

    private func setupEventHandler() {
        var eventTypes = [
            EventTypeSpec(eventClass: OSType(kEventClassKeyboard), eventKind: UInt32(kEventHotKeyPressed)),
            EventTypeSpec(eventClass: OSType(kEventClassKeyboard), eventKind: UInt32(kEventHotKeyReleased))
        ]

        let ptr = UnsafeMutablePointer<HotkeyManager>.allocate(capacity: 1)
        unsafe ptr.initialize(to: self)
        unsafe contextPtr = ptr

        let status = unsafe InstallEventHandler(
            GetEventDispatcherTarget(),
            { _, theEvent, userData -> OSStatus in
                guard let userData = unsafe userData else { return OSStatus(eventNotHandledErr) }
                let manager = unsafe userData.assumingMemoryBound(to: HotkeyManager.self).pointee

                var hotKeyID = EventHotKeyID()
                let getStatus = unsafe GetEventParameter(
                    theEvent,
                    UInt32(kEventParamDirectObject),
                    UInt32(typeEventHotKeyID),
                    nil,
                    MemoryLayout<EventHotKeyID>.size,
                    nil,
                    &hotKeyID
                )
                guard getStatus == noErr else { return OSStatus(eventNotHandledErr) }

                let eventKind = unsafe GetEventKind(theEvent)
                manager.dispatch(hotkeyID: hotKeyID.id, eventKind: eventKind)
                return noErr
            },
            eventTypes.count,
            &eventTypes,
            ptr,
            &eventHandler
        )

        if status != noErr {
            AKError("hs.hotkey: Failed to install Carbon event handler (error \(status))")
        }
    }

    func register(hotkeyID: UInt32, hotkey: HSHotkey) {
        hotkeys[hotkeyID] = hotkey
    }

    func unregister(hotkeyID: UInt32) {
        hotkeys.removeValue(forKey: hotkeyID)
    }

    private func dispatch(hotkeyID: UInt32, eventKind: UInt32) {
        hotkeys[hotkeyID]?.trigger(eventKind: eventKind)
    }
}

// MARK: - FourCharCode helper

extension NSString {
    var fourCharCode: FourCharCode {
        guard self.length == 4 else { return 0 }
        var result: FourCharCode = 0
        for i in 0..<4 {
            result = (result << 8) + FourCharCode(self.character(at: i))
        }
        return result
    }
}

File: Hammerspoon 2/Modules/hs.hotkey/hs.hotkey.js
"use strict";

/// Bind a hotkey from a single options object. Like hs.hotkey.bind()/create(), but also
/// accepts a `message` and a `repeat` callback. `message` is available on any hotkey (not
/// just ones created via bindSpec()) by setting `.message` directly on the returned object;
/// see hs.hotkey's message property for exactly when it is shown.
/// Parameters:
///  - spec: An object with the following fields:
///    - mods: {string[]} Modifier keys (e.g. ["cmd", "shift"])
///    - key: {string} Key name or character (e.g. "h")
///    - message: {string} [optional] A description shown as a toast when the hotkey fires
///    - pressed: {(() => void) | null} [optional] Called when the hotkey is pressed
///    - released: {(() => void) | null} [optional] Called when the hotkey is released
///    - repeat: {(() => void) | null} [optional] Called repeatedly while the hotkey is held down
/// Returns: A hotkey object, or null if binding failed
/// Example:
/// ```js
/// hs.hotkey.bindSpec({
///     mods: ["cmd"], key: "space", message: "Spotlight-like",
///     pressed: () => console.log("pressed")
/// })
/// ```
hs.hotkey.bindSpec = function(spec) {
    const { mods = [], key, message, pressed = null, released = null, repeat: repeatfn = null } = spec;
    const hk = hs.hotkey.bind(mods, key, pressed, released);
    if (!hk) return null;
    if (message) hk.message = message;
    if (repeatfn) hk.callbackRepeat = repeatfn;
    return hk;
};

/**
 * A modal hotkey group returned by hs.hotkey.createModal(). Hotkeys bound to the modal
 * via bind() are only enabled while the modal is active (i.e. between enter() and exit()).
 * @property {boolean} isActive - Whether the modal is currently active
 * @property {Function|null} enterFn - Callback invoked when the modal is entered
 * @property {Function|null} exitFn - Callback invoked when the modal is exited
 */
class HSHotkeyModal {
    constructor() {
        this._hotkeys = [];
        this._trigger = null;
        this.enterFn = null;
        this.exitFn = null;
        this.isActive = false;
    }

    /**
     * Bind a hotkey to this modal. The hotkey is only enabled while the modal is active.
     * @param {string[]} mods - Modifier keys for the hotkey (e.g. ["cmd", "shift"])
     * @param {string} key - Key name for the hotkey (e.g. "h")
     * @param {Function|null} callbackPressed - Called when the hotkey is pressed, or null
     * @param {Function|null} callbackReleased - Called when the hotkey is released, or null
     * @param {Function|null} [callbackRepeat] - Called repeatedly while the hotkey is held down, or null
     * @returns {HSHotkeyModal} This modal, for chaining
     */
    bind(mods, key, callbackPressed, callbackReleased, callbackRepeat) {
        const hk = hs.hotkey.create(mods, key, callbackPressed, callbackReleased);
        if (!hk) return this;
        if (callbackRepeat) hk.callbackRepeat = callbackRepeat;
        this._hotkeys.push(hk);
        if (this.isActive) hk.enable();
        return this;
    }

    /**
     * Enter the modal: its trigger (if any) is disabled and its bound hotkeys are enabled.
     * @returns {HSHotkeyModal} This modal, for chaining
     */
    enter() {
        if (this.isActive) return this;
        this.isActive = true;
        if (this._trigger) this._trigger.disable();
        for (const hk of this._hotkeys) hk.enable();
        if (typeof this.enterFn === 'function') {
            try { this.enterFn(); } catch(e) { console.error("hs.hotkey modal enterFn error: " + e); }
        }
        console.debug("hs.hotkey: modal entered")
        return this;
    }

    /**
     * Exit the modal: its bound hotkeys are disabled and its trigger (if any) is re-enabled.
     * @returns {HSHotkeyModal} This modal, for chaining
     */
    exit() {
        if (!this.isActive) return this;
        this.isActive = false;
        for (const hk of this._hotkeys) hk.disable();
        if (this._trigger) this._trigger.enable();
        if (typeof this.exitFn === 'function') {
            try { this.exitFn(); } catch(e) { console.error("hs.hotkey modal exitFn error: " + e); }
        }
        console.debug("hs.hotkey: modal exited")
        return this;
    }

    /**
     * Destroy the modal, along with its trigger and all hotkeys bound to it.
     */
    destroy() {
        this.isActive = false;
        for (const hk of this._hotkeys) hk.destroy();
        this._hotkeys = [];
        if (this._trigger) {
            this._trigger.destroy();
            this._trigger = null;
        }
    }
}

/// Create a new modal hotkey group, optionally entered via a trigger key combination
/// Parameters:
///  - mods: Modifier keys for the trigger hotkey (e.g. ["cmd", "shift"]), or an empty array for no trigger
///  - key: Key name for the trigger hotkey (e.g. "h"), or an empty string for no trigger
/// Returns: {HSHotkeyModal} A modal object with bind(), enter(), exit(), destroy() methods, isActive property, and enterFn/exitFn callbacks
/// Example:
/// ```js
/// const m = hs.hotkey.createModal(['cmd'], 'h')
/// m.bind(['shift'], 'j', () => console.log('shift-j pressed'), null)
/// m.enterFn = () => console.log('modal entered')
/// m.exitFn  = () => console.log('modal exited')
/// m.bind([], 'escape', () => m.exit(), null)
/// ```
hs.hotkey.createModal = function(mods, key) {
    const modal = new HSHotkeyModal();

    if (key !== '') {
        const trigger = hs.hotkey.bind(mods, key, () => modal.enter(), null);
        if (trigger) modal._trigger = trigger;
    }

    return modal;
};

/// Create and enable a hotkey that, while held down, displays a list of all currently
/// enabled hotkeys (and their messages, if any) as an on-screen toast.
/// Parameters:
///  - mods: Modifier keys for the trigger hotkey (e.g. ["cmd", "shift"])
///  - key: Key name for the trigger hotkey (e.g. "/")
/// Returns: A hotkey object
/// Example:
/// ```js
/// hs.hotkey.showHotkeys(["cmd","shift"], "/")
/// ```
hs.hotkey.showHotkeys = function(mods, key) {
    let alert = null;
    return hs.hotkey.bind(mods, key, () => {
        const lines = hs.hotkey.getHotkeys().map(hk => {
            const label = hk.mods.join('+') + (hk.mods.length ? '+' : '') + hk.key;
            return hk.message ? `${label}: ${hk.message}` : label;
        });
        alert = hs.ui.alert(lines.join('\n')).duration(3600).show();
    }, () => {
        if (alert) {
            alert.close();
            alert = null;
        }
    });
};
